File: pyagentspec/src/pyagentspec/adapters/langgraph/agentspecexporter.py
# Copyright © 2025 Oracle and/or its affiliates.
#
# This software is under the Apache License 2.0
# (LICENSE-APACHE or http://www.apache.org/licenses/LICENSE-2.0) or Universal Permissive License
# (UPL) 1.0 (LICENSE-UPL or https://oss.oracle.com/licenses/upl), at your option.

from typing import Any, Literal, Optional, Tuple, Union, overload

from pyagentspec.adapters._agentspecexporter import (
    AdapterAgnosticAgentSpecExporter,
    _RuntimeDisaggregatedComponentsConfigT,
)
from pyagentspec.adapters._agentspecloader import RuntimeToAgentSpecConverter, _RuntimeComponentT
from pyagentspec.adapters.langgraph._agentspecconverter import LangGraphToAgentSpecConverter
from pyagentspec.versioning import AgentSpecVersionEnum


class AgentSpecExporter(AdapterAgnosticAgentSpecExporter):
    """Helper class to convert LangGraph objects into Agent Spec configuration."""

    @property
    def runtime_to_agentspec_converter(self) -> RuntimeToAgentSpecConverter:
        return LangGraphToAgentSpecConverter()

    @overload
    def to_json(self, runtime_component: _RuntimeComponentT) -> str: ...

    @overload
    def to_json(
        self,
        runtime_component: _RuntimeComponentT,
        agentspec_version: Optional[AgentSpecVersionEnum],
    ) -> str: ...

    @overload
    def to_json(
        self,
        runtime_component: _RuntimeComponentT,
        *,
        disaggregated_components: _RuntimeDisaggregatedComponentsConfigT,
        export_disaggregated_components: Literal[True],
    ) -> Tuple[str, str]: ...

    @overload
    def to_json(
        self,
        runtime_component: _RuntimeComponentT,
        *,
        disaggregated_components: Optional[_RuntimeDisaggregatedComponentsConfigT],
        export_disaggregated_components: Literal[False],
    ) -> str: ...

    @overload
    def to_json(
        self,
        runtime_component: _RuntimeComponentT,
        *,
        disaggregated_components: Optional[_RuntimeDisaggregatedComponentsConfigT],
        export_disaggregated_components: bool,
    ) -> Union[str, Tuple[str, str]]: ...

    @overload
    def to_json(
        self,
        runtime_component: _RuntimeComponentT,
        agentspec_version: Optional[AgentSpecVersionEnum],
        disaggregated_components: Optional[_RuntimeDisaggregatedComponentsConfigT],
        export_disaggregated_components: bool,
    ) -> Union[str, Tuple[str, str]]: ...

    def to_json(
        self,
        runtime_component: _RuntimeComponentT,
        agentspec_version: Optional[AgentSpecVersionEnum] = None,
        disaggregated_components: Optional[_RuntimeDisaggregatedComponentsConfigT] = None,
        export_disaggregated_components: bool = False,
    ) -> Union[str, Tuple[str, str]]:
        """
        Transform the given LangGraph component into the respective Agent Spec JSON representation.

        Parameters
        ----------
        runtime_component:
            LangGraph component to serialize to an Agent Spec configuration.
        agentspec_version:
            The Agent Spec version of the component.
        disaggregated_components:
            Configuration specifying the components/fields to disaggregate upon serialization.
            Each item can be:

            - A ``Component``: to disaggregate the component using its id
            - A tuple ``(Component, str)``: to disaggregate the component using
              a custom id.

            .. note::

                Components in ``disaggregated_components`` are disaggregated
                even if ``export_disaggregated_components`` is ``False``.
        export_disaggregated_components:
            Whether to export the disaggregated components or not. Defaults to ``False``.

        Returns
        -------
        If ``export_disaggregated_components`` is ``True``:

        str
            The JSON serialization of the root component.
        str
            The JSON serialization of the disaggregated components.

        If ``export_disaggregated_components`` is ``False``:

        str
            The JSON serialization of the root component.
        """
        return super().to_json(
            runtime_component,
            agentspec_version=agentspec_version,
            disaggregated_components=disaggregated_components,
            export_disaggregated_components=export_disaggregated_components,
        )

    @overload
    def to_yaml(self, runtime_component: _RuntimeComponentT) -> str: ...

    @overload
    def to_yaml(
        self,
        runtime_component: _RuntimeComponentT,
        agentspec_version: Optional[AgentSpecVersionEnum],
    ) -> str: ...

    @overload
    def to_yaml(
        self,
        runtime_component: _RuntimeComponentT,
        *,
        disaggregated_components: _RuntimeDisaggregatedComponentsConfigT,
        export_disaggregated_components: Literal[True],
    ) -> Tuple[str, str]: ...

    @overload
    def to_yaml(
        self,
        runtime_component: _RuntimeComponentT,
        *,
        disaggregated_components: Optional[_RuntimeDisaggregatedComponentsConfigT],
        export_disaggregated_components: Literal[False],
    ) -> str: ...

    @overload
    def to_yaml(
        self,
        runtime_component: _RuntimeComponentT,
        *,
        disaggregated_components: Optional[_RuntimeDisaggregatedComponentsConfigT],
        export_disaggregated_components: bool,
    ) -> Union[str, Tuple[str, str]]: ...

    @overload
    def to_yaml(
        self,
        runtime_component: _RuntimeComponentT,
        agentspec_version: Optional[AgentSpecVersionEnum],
        disaggregated_components: Optional[_RuntimeDisaggregatedComponentsConfigT],
        export_disaggregated_components: bool,
    ) -> Union[str, Tuple[str, str]]: ...

    def to_yaml(
        self,
        runtime_component: _RuntimeComponentT,
        agentspec_version: Optional[AgentSpecVersionEnum] = None,
        disaggregated_components: Optional[_RuntimeDisaggregatedComponentsConfigT] = None,
        export_disaggregated_components: bool = False,
    ) -> Union[str, Tuple[str, str]]:
        """
        Transform the given LangGraph component into the respective Agent Spec YAML representation.

        Parameters
        ----------
        runtime_component:
            LangGraph component to serialize to an Agent Spec configuration.
        agentspec_version:
            The Agent Spec version of the component.
        disaggregated_components:
            Configuration specifying the components/fields to disaggregate upon serialization.
            Each item can be:

            - A ``Component``: to disaggregate the component using its id
            - A tuple ``(Component, str)``: to disaggregate the component using
              a custom id.

            .. note::

                Components in ``disaggregated_components`` are disaggregated
                even if ``export_disaggregated_components`` is ``False``.
        export_disaggregated_components:
            Whether to export the disaggregated components or not. Defaults to ``False``.

        Returns
        -------
        If ``export_disaggregated_components`` is ``True``:

        str
            The YAML serialization of the root component.
        str
            The YAML serialization of the disaggregated components.

        If ``export_disaggregated_components`` is ``False``:

        str
            The YAML serialization of the root component.
        """
        return super().to_yaml(
            runtime_component,
            agentspec_version=agentspec_version,
            disaggregated_components=disaggregated_components,
            export_disaggregated_components=export_disaggregated_components,
        )

    @overload
    def to_dict(self, runtime_component: _RuntimeComponentT) -> dict[str, Any]: ...

    @overload
    def to_dict(
        self,
        runtime_component: _RuntimeComponentT,
        agentspec_version: Optional[AgentSpecVersionEnum],
    ) -> dict[str, Any]: ...

    @overload
    def to_dict(
        self,
        runtime_component: _RuntimeComponentT,
        *,
        disaggregated_components: _RuntimeDisaggregatedComponentsConfigT,
        export_disaggregated_components: Literal[True],
    ) -> Tuple[str, dict[str, Any]]: ...

    @overload
    def to_dict(
        self,
        runtime_component: _RuntimeComponentT,
        *,
        disaggregated_components: Optional[_RuntimeDisaggregatedComponentsConfigT],
        export_disaggregated_components: Literal[False],
    ) -> dict[str, Any]: ...

    @overload
    def to_dict(
        self,
        runtime_component: _RuntimeComponentT,
        *,
        disaggregated_components: Optional[_RuntimeDisaggregatedComponentsConfigT],
        export_disaggregated_components: bool,
    ) -> Union[dict[str, Any], Tuple[str, dict[str, Any]]]: ...

    @overload
    def to_dict(
        self,
        runtime_component: _RuntimeComponentT,
        agentspec_version: Optional[AgentSpecVersionEnum],
        disaggregated_components: Optional[_RuntimeDisaggregatedComponentsConfigT],
        export_disaggregated_components: bool,
    ) -> Union[dict[str, Any], Tuple[str, dict[str, Any]]]: ...

    def to_dict(
        self,
        runtime_component: _RuntimeComponentT,
        agentspec_version: Optional[AgentSpecVersionEnum] = None,
        disaggregated_components: Optional[_RuntimeDisaggregatedComponentsConfigT] = None,
        export_disaggregated_components: bool = False,
    ) -> Union[dict[str, Any], Tuple[str, dict[str, Any]]]:
        """
        Transform the given LangGraph component into the respective Agent Spec dictionary.

        Parameters
        ----------
        runtime_component:
            LangGraph component to serialize to an Agent Spec configuration.
        agentspec_version:
            The Agent Spec version of the component.
        disaggregated_components:
            Configuration specifying the components/fields to disaggregate upon serialization.
            Each item can be:

            - A ``Component``: to disaggregate the component using its id
            - A tuple ``(Component, str)``: to disaggregate the component using
              a custom id.

            .. note::

                Components in ``disaggregated_components`` are disaggregated
                even if ``export_disaggregated_components`` is ``False``.
        export_disaggregated_components:
            Whether to export the disaggregated components or not. Defaults to ``False``.

        Returns
        -------
        If ``export_disaggregated_components`` is ``True``:

        str
            The dictionary serialization of the root component.
        str
            The dictionary serialization of the disaggregated components.

        If ``export_disaggregated_components`` is ``False``:

        str
            The dictionary serialization of the root component.
        """
        return super().to_dict(
            runtime_component,
            agentspec_version=agentspec_version,
            disaggregated_components=disaggregated_components,
            export_disaggregated_components=export_disaggregated_components,
        )

File: tsagentspec/src/adapters/common/agentspec-exporter.ts
/**
 * Framework-agnostic exporter converting runtime objects to Agent Spec
 * configurations. Port of
 * `pyagentspec.adapters._agentspecexporter.AdapterAgnosticAgentSpecExporter`.
 */
import type { ComponentBase } from "../../component.js";
import { AgentSpecSerializer } from "../../serialization/index.js";
import type {
  ComponentSerializationPlugin,
  DisaggregatedComponentsConfig,
} from "../../serialization/index.js";
import type { AgentSpecVersion } from "../../versioning.js";
import type { RuntimeToAgentSpecConverter } from "./converters.js";

/**
 * Runtime components/fields to disaggregate upon serialization. Each item can
 * be a runtime component (disaggregated using its converted component id) or
 * a `[runtimeComponent, customId]` pair (disaggregated using the custom id).
 */
export type RuntimeDisaggregatedComponentsConfig = ReadonlyArray<
  unknown | readonly [unknown, string]
>;

/** Options for the export methods. */
export interface ExportOptions {
  /** The Agent Spec version to serialize the component at. */
  agentspecVersion?: AgentSpecVersion;
  /**
   * Components/fields to disaggregate upon serialization. Components listed
   * here are disaggregated even if `exportDisaggregatedComponents` is false.
   */
  disaggregatedComponents?: RuntimeDisaggregatedComponentsConfig;
  /** Whether to export the disaggregated components. Defaults to false. */
  exportDisaggregatedComponents?: boolean;
}

/** Serialized dictionary form of a component. */
export type ExportedDict = Record<string, unknown>;

/** Helper class to convert runtime objects to Agent Spec configurations. */
export abstract class AdapterAgnosticAgentSpecExporter {
  /** Serialization plugins passed to the serializer. */
  readonly plugins: ComponentSerializationPlugin[];

  constructor(plugins?: ComponentSerializationPlugin[]) {
    this.plugins = plugins ?? [];
  }

  /** Converter used to convert runtime components to Agent Spec components. */
  abstract get runtimeToAgentSpecConverter(): RuntimeToAgentSpecConverter;

  /**
   * Transform the given runtime component into the respective Agent Spec JSON
   * representation. Returns `[main, referenced]` JSON strings when
   * `exportDisaggregatedComponents` is true.
   */
  toJson(
    runtimeComponent: unknown,
    options?: ExportOptions,
  ): string | [string, string] {
    return this._export(
      (serializer, agentspecAssistant, serializerOptions) =>
        serializer.toJson(agentspecAssistant, serializerOptions),
      runtimeComponent,
      options,
    );
  }

  /**
   * Transform the given runtime component into the respective Agent Spec YAML
   * representation. Returns `[main, referenced]` YAML strings when
   * `exportDisaggregatedComponents` is true.
   */
  toYaml(
    runtimeComponent: unknown,
    options?: ExportOptions,
  ): string | [string, string] {
    return this._export(
      (serializer, agentspecAssistant, serializerOptions) =>
        serializer.toYaml(agentspecAssistant, serializerOptions),
      runtimeComponent,
      options,
    );
  }

  /**
   * Transform the given runtime component into the respective Agent Spec
   * dictionary. Returns `[main, referenced]` dictionaries when
   * `exportDisaggregatedComponents` is true.
   */
  toDict(
    runtimeComponent: unknown,
    options?: ExportOptions,
  ): ExportedDict | [ExportedDict, ExportedDict] {
    return this._export(
      (
        serializer,
        agentspecAssistant,
        serializerOptions,
      ): ExportedDict | [ExportedDict, ExportedDict] => {
        // The TS AgentSpecSerializer has no public toDict, so the dictionary
        // form is derived from the JSON serialization.
        const json = serializer.toJson(agentspecAssistant, serializerOptions);
        if (Array.isArray(json)) {
          return [
            JSON.parse(json[0]) as ExportedDict,
            JSON.parse(json[1]) as ExportedDict,
          ];
        }
        return JSON.parse(json) as ExportedDict;
      },
      runtimeComponent,
      options,
    );
  }

  /**
   * Transform the given runtime component into the respective AgentSpec
   * component.
   */
  toComponent(runtimeComponent: unknown): ComponentBase {
    return this.runtimeToAgentSpecConverter.convert(runtimeComponent);
  }

  /**
   * Common implementation of the export methods. Each public method passes
   * the closure that serializes the converted component to its output form.
   */
  protected _export<SerializedT>(
    serialize: (
      serializer: AgentSpecSerializer,
      agentspecAssistant: ComponentBase,
      serializerOptions: {
        agentspecVersion?: AgentSpecVersion;
        disaggregatedComponents?: DisaggregatedComponentsConfig;
        exportDisaggregatedComponents: boolean;
      },
    ) => SerializedT,
    runtimeComponent: unknown,
    options?: ExportOptions,
  ): SerializedT {
    const serializer = new AgentSpecSerializer(this.plugins);

    const [convertedDisagComponents, referencedComponents] =
      options?.disaggregatedComponents !== undefined
        ? this._convertDisaggregatedConfig(options.disaggregatedComponents)
        : [undefined, undefined];
    const agentspecAssistant = this.runtimeToAgentSpecConverter.convert(
      runtimeComponent,
      referencedComponents,
    );
    const serializerOptions = {
      agentspecVersion: options?.agentspecVersion,
      disaggregatedComponents: convertedDisagComponents,
      exportDisaggregatedComponents:
        options?.exportDisaggregatedComponents ?? false,
    };
    return serialize(serializer, agentspecAssistant, serializerOptions);
  }

  /**
   * Convert the runtime disaggregated-components config into Agent Spec
   * components, accumulating the shared referenced-objects registry that is
   * then passed into the root conversion (so disaggregated components share
   * identity with references inside the root component).
   */
  protected _convertDisaggregatedConfig(
    runtimeDisagConfig: RuntimeDisaggregatedComponentsConfig,
  ): [DisaggregatedComponentsConfig, Map<string, ComponentBase>] {
    const agentspecDisaggregatedComponents: Array<
      ComponentBase | readonly [ComponentBase, string]
    > = [];
    const referencedComponents = new Map<string, ComponentBase>();
    for (const disagConfig of runtimeDisagConfig) {
      const pair =
        Array.isArray(disagConfig) &&
        disagConfig.length === 2 &&
        typeof disagConfig[1] === "string"
          ? (disagConfig as unknown as readonly [unknown, string])
          : undefined;
      const runtimeComponent = pair !== undefined ? pair[0] : disagConfig;
      const agentspecComponent = this.runtimeToAgentSpecConverter.convert(
        runtimeComponent,
        referencedComponents,
      );
      // Mirroring Python, the converted component keeps its own id everywhere
      // (root tree and disaggregated registry); a custom id is passed through
      // as a [component, customId] pair and applied by the serializer only as
      // the serialization-time mapping key.
      agentspecDisaggregatedComponents.push(
        pair !== undefined
          ? ([agentspecComponent, pair[1]] as const)
          : agentspecComponent,
      );
    }
    return [agentspecDisaggregatedComponents, referencedComponents];
  }
}
